sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/KafkaCosmosWriterBase.java

```diff
@@ -0,0 +1,121 @@
+        List<SinkOperation> sinkOperations =
+            sinkRecords
+                .stream()
+                .map(sinkRecord -> new SinkOperation(sinkRecord))
+                .collect(Collectors.toList());
+
+        try {
+            writeCore(container, sinkOperations);
+        } catch (Exception e) {
+            LOGGER.error("Write failed. ", e);
+            throw new CosmosDBWriteException(e.getMessage());
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    protected String getId(Object recordValue) {
+        checkArgument(recordValue instanceof Map, "Argument 'recordValue' is not valid map format.");
+        return ((Map<String, Object>) recordValue).get(ID).toString();
+    }
+
+    @SuppressWarnings("unchecked")
+    protected String getEtag(Object recordValue) {
+        checkArgument(recordValue instanceof Map, "Argument 'recordValue' is not valid map format.");
+        return ((Map<String, Object>) recordValue).getOrDefault(ETAG, Strings.Emtpy).toString();
+    }
+
+    @SuppressWarnings("unchecked")
+    protected PartitionKey getPartitionKeyValue(Object recordValue, PartitionKeyDefinition partitionKeyDefinition) {
+        checkArgument(recordValue instanceof Map, "Argument 'recordValue' is not valid map format.");
+
+        //TODO[Public Preview]: add support for sub-partition
+        String partitionKeyPath = StringUtils.join(partitionKeyDefinition.getPaths(), "");
+        Map<String, Object> recordMap = (Map<String, Object>) recordValue;
+        Object partitionKeyValue = recordMap.get(partitionKeyPath.substring(1));
+        PartitionKeyInternal partitionKeyInternal = PartitionKeyInternal.fromObjectArray(Collections.singletonList(partitionKeyValue), false);
+
+        return ImplementationBridgeHelpers
+            .PartitionKeyHelper
+            .getPartitionKeyAccessor()
+            .toPartitionKey(partitionKeyInternal);
+    }
+
+    protected boolean shouldRetry(Throwable exception, int attemptedCount, int maxRetryCount) {
+        if (attemptedCount >= maxRetryCount) {
+            return false;
+        }
+
+        return KafkaCosmosExceptionsHelper.isTransientFailure(exception);
+    }
+
+    protected Mono<PartitionKeyDefinition> getPartitionKeyDefinition(CosmosAsyncContainer container) {
+        return Mono.just(CosmosBridgeInternal.getAsyncDocumentClient(container.getDatabase()).getCollectionCache())
+            .flatMap(collectionCache -> {
+                return collectionCache
+                    .resolveByNameAsync(
+                        null,
+                        ImplementationBridgeHelpers
+                            .CosmosAsyncContainerHelper
+                            .getCosmosAsyncContainerAccessor()
+                            .getLinkWithoutTrailingSlash(container),
+                        null,
+                        new DocumentCollection())
```

Comment:
why new DocumentCollection and not null? I think his param is meant to be used when asynccache has an instance believed to be stale?

Reply:
both works. The value used here just mean if the cached value is not the same as the staled value, then use the cached value directly. So null or new DocumentCollection() would have the same effect

Reply:
BUt null is cheaper (no new garbage instantiation)

Reply:
will change in next PR